contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,312 @@
+            Self::owner(&env).require_auth();
+
+            env.deployer().update_current_contract_wasm(new_wasm_hash);
+            Ok(())
+        }
+    }
+
+    #[contractimpl]
+    impl DummyContract {
+        pub fn __constructor(env: Env, owner: Address) {
+            env.storage().instance().set(&DataKey::Owner, &owner)
+        }
+
+        fn owner(env: &Env) -> Address {
+            env.storage().instance().get(&DataKey::Owner).unwrap()
+        }
+    }
+
+    #[contracttype]
+    enum DataKey {
+        Data,
+        Owner,
+    }
+
+    #[contracterror]
+    enum ContractError {
+        SomeFailure = 1,
+    }
+
+    /// Dummy contract logic after upgrade is loaded into WASM_AFTER_UPGRADE
+    ///
+    /// #[contractimpl]
+    /// impl UpgradeableInterface for DummyContract {
+    ///     type Error = ContractError;
+    ///
+    ///     fn version(env: Env) -> String {
+    ///         String::from_str(&env, "0.2.0")
+    ///     }
+    ///
+    ///     fn upgrade(env: Env, new_wasm_hash: BytesN<32>) -> Result<(), ContractError> {
+    ///         Self::owner(&env).require_auth();
+    ///
+    ///         env.deployer().update_current_contract_wasm(new_wasm_hash);
+    ///         Ok(())
+    ///     }
+    /// }
+    ///
+    /// #[contractimpl]
+    /// impl DummyContract {
+    ///     pub fn migrate(env: Env, migration_data: String) {
+    ///         Self::owner(&env).require_auth();
+    ///
+    ///         env.storage()
+    ///             .instance()
+    ///             .set(&DataKey::Data, &migration_data);
+    ///     }
+    ///
+    ///     fn owner(env: &Env) -> Address {
+    ///         env.storage().instance().get(&DataKey::Owner).unwrap()
+    ///     }
+    /// }
+    const WASM_AFTER_UPGRADE: &[u8] = include_bytes!("testdata/dummy.wasm");
+
+    #[test]
+    fn upgrade_and_migrate_are_atomic() {
+        let env = Env::default();
+
+        let owner = Address::generate(&env);
+        let contract_address = env.register(DummyContract, (&owner,));
+        let upgrader_address = env.register(Upgrader, ());
+
+        let original_version: String = query_version(&env, &contract_address);
+        assert_eq!(original_version, String::from_str(&env, "0.1.0"));
+
+        let hash_after_upgrade = env.deployer().upload_contract_wasm(WASM_AFTER_UPGRADE);
+        let expected_data = String::from_str(&env, "migration successful");
+        let expected_version = String::from_str(&env, "0.2.0");
+
+        let (upgrade_auth, migrate_auth) =
+            build_invocation_auths(&env, &contract_address, &hash_after_upgrade, &expected_data);
+
+        // add the owner to the set of authenticated addresses
+        env.mock_auths(&[
+            MockAuth {
+                address: &owner,
+                invoke: &upgrade_auth,
+            },
+            MockAuth {
+                address: &owner,
+                invoke: &migrate_auth,
```

Comment:
Try using `assert_invoke_auth_err` instead? You can add another pattern for it where it takes a list of callers and invocation calls and creates mock auths for them, and then calls the final method to test the error?

Reply:
the auth that I need to mock here is very specific for this scenario: I call function A, but need to authenticate functions B and C that have very specific subsets of the input parameters. Also, the tests are modulating what is authenticated and what is not, so there is no good macro solution, and it wouldn't be reusable anyway